test/adaptive_avg_pool/problem.py

```diff
@@ -0,0 +1,186 @@
+        self.works = works
+        self.device = device
+        self.seed = seed
+        self.id_prefix = id_prefix
+
+    def make_id(self) -> str:
+        """Create an id from problem parameters.
+
+        Returns:
+            problem id
+        """
+        prefix = (self.id_prefix + "-") if self.id_prefix != "" else ""
+        return (
+            prefix + f"dev={self.device}-N={self.N}-in={self.shape_input}-"
+            f"out={self.shape_target}-works={self.works}"
+        )
+
+    def set_up(self) -> None:
+        """Set up problem and do one forward pass."""
+        torch.manual_seed(self.seed)
+        self.module = self._make_module()
+        self.input = randn(self.shape_input)
+        self.output = self.module(self.input)
+
+    def tear_down(self):
+        """Delete created torch variables."""
+        del self.module
+        del self.input
+        del self.output
+
+    def _make_module(self) -> Union[AdaptiveAvgPool1d, AdaptiveAvgPool2d]:
+        if self.N == 1:
+            module = AdaptiveAvgPool1d(output_size=self.shape_target)
+        elif self.N == 2:
+            module = AdaptiveAvgPool2d(output_size=self.shape_target)
+        elif self.N == 3:
+            module = AdaptiveAvgPool3d(output_size=self.shape_target)
+        else:
+            raise NotImplementedError(f"N={self.N} not implemented in test suite.")
+        return extend(module.to(device=self.device))
+
+    def check_parameters(self) -> None:
+        """Key method for test.
+
+        Run the AdaptiveAvgPoolNDDerivatives.check_parameters() method.
+        """
+        self._get_derivatives().check_parameters(module=self.module)
+
+    def _get_derivatives(self) -> AdaptiveAvgPoolNDDerivatives:
+        if self.N in [1, 2, 3]:
+            derivatives = AdaptiveAvgPoolNDDerivatives(N=self.N)
+        else:
+            raise NotImplementedError(f"N={self.N} not implemented in test suite.")
+        return derivatives
+
+    def check_equivalence(self) -> None:
+        """Check if the given parameters lead to the same output.
+
+        Checks the sizes and values.
+        """
+        stride, kernel_size, _ = self._get_derivatives().get_parameters(self.module)
+        module_equivalent: Module = self._make_module_equivalent(stride, kernel_size)
+        output_equivalent: Tensor = module_equivalent(self.input)
+
+        check_sizes_and_values(self.output, output_equivalent)
+
+    def _make_module_equivalent(
+        self, stride: List[int], kernel_size: List[int]
+    ) -> Union[AvgPool1d, AvgPool2d]:
+        if self.N == 1:
+            module = AvgPool1d(kernel_size=kernel_size[0], stride=stride[0])
+        elif self.N == 2:
+            module = AvgPool2d(kernel_size=kernel_size, stride=stride)
+        elif self.N == 3:
+            module = AvgPool3d(kernel_size=kernel_size, stride=stride)
+        else:
+            raise NotImplementedError(f"N={self.N} not implemented in test suite.")
```

Comment:
move check to constructor

Reply:
Yes. However, I am leaving a simple `raise NotImplementedError` because otherwise checkers are complaining that `Local variable 'module' might be referenced before assignment`

Reply:
How about
```
pooling_cls = {1: AvgPool, ...}
module = pooling_cls[self.N](kernel_size, ...)
```
which avoids the assignment problem?

Reply:
Yes. (I never used this pattern but makes a lot of sense)